contributors/design-proposals/device-plugin.md

```diff
@@ -0,0 +1,794 @@
+	// Minimum version the Kubelet API supports.
+	string version = 1;
+	// Kubelet fills this field if it encounters any errors
+	// during the registration process or discover process
+	string error = 2;
+}
+
+message HeartbeatRequest {
+	string vendor = 1;
+}
+
+message HeartbeatResponse {
+	// Kubelet answers with a string telling the device
+	// plugin to either re-register itself or not
+	string response = 1;
+	// Kubelet fills this field if it encountered any errors
+	string error = 2;
+}
+
+message AllocateRequest {
+	repeated Device devices = 1;
+}
+
+message AllocateResponse {
+	// List of environment variable to set in the container.
+	repeated KeyValue envs = 1;
+	// Mounts for the container.
+	repeated Mount mounts = 2;
+}
+
+// E.g:
+// struct Device {
+//    Kind: "NVIDIA-gpu"
+//    Name: "GPU-fef8089b-4820-abfc-e83e-94318197576e",
+//    Health: "Healthy",
+//    Properties: {
+//        "Family": "Pascal",
+//        "Memory": "4G",
+//        "ECC"   : "True",
+//    }
+//}
+message Device {
+	string Kind = 1;
+	string Name = 2;
+	string Health = 3;
+	string Vendor = 4;
+	map<string, string> properties = 5; // Could be [1, 1.2, 1G]
+}
+```
+
+### HealthCheck and Failure Recovery
+
+We want Kubelet as well as the Device Plugins to recover from failures
+that may happen on any side of this protocol.
+
+At the communication level, gRPC is a very strong piece of software and
+is able to ensure that if failure happens it will try it's best to recover
+through exponential backoff reconnection.
+
+The proposed mechanism intends to replace any device specific handling in
+Kubelet. Therefore in general, device plugin failure or upgrade means that
+Kubelet is not able to accept any pod requesting a Device until the upgrade
+or failure finishes.
+
+If a device fails, the Device Plugin should signal that through the HealthCheck
+stream and we expect Kubelet to stop the pod and reschedule it.
+
+If any Device Plugin fails the behavior we expect depends on the task Kubelet
+is performing:
+* In general we expect Kubelet to remove any devices that are owned by the failed
+  device plugin from the resources advertised by the Node status.
+* We however do not expect Kubelet to fail or restart any pods or containers
+  running that are using these devices.
+* If Kubelet is in the process of allocating a device, then it should fail
+  the container process and reschedule the Pod.
+
+If the Kubelet fails or restarts, we expect the Device Plugins to know about
+it through Kubelet's Heartbeat call which every Device Plugin should call
+every 5s.
+
+When Kubelet fails or restarts it should know what are the devices that are
+owned by the different containers and be able to rebuild a list of available
+devices.
+In the current design, instead of checkpointing this data, we propose to save
+this in the API server as this gives introspection capabilities to the user
+has minimal impact on performances and is a minimal change that can be
```

Comment:
Checkpointing to API server is inherently racy. Kubelet's writes to API server is rate limited and so it is handled asynchronously. It is not a good idea to keep exposing all kubelet internals via the API just because it's convenient.